include/vata2/simutil/binary_relation.hh

```diff
@@ -0,0 +1,935 @@
+		std::vector<bool>::iterator dst = tmp.begin();
+		for (size_t i = 0; i < size_; ++i) {
+			std::copy(src, src + size_, dst);
+			src += rowSize_;
+			dst += newRowSize;
+		}
+		std::swap(data_, tmp);
+		rowSize_ = newRowSize;
+	}
+
+	void grow(size_t newSize, bool defVal = false)
+	{
+		assert(rowSize_ <= newSize);
+
+		size_t newRowSize = rowSize_;
+		while (newRowSize <= newSize)
+			newRowSize <<= 1;
+		assert(newSize <= newRowSize);
+		this->realloc(newRowSize, defVal);
+
+	}
+/*
+	void shrinkToFit() {
+
+		assert(this->row_size_ > this->size_);
+		size_t newRowSize = this->row_size_;
+		while (this->size_ < (newRowSize >> 1))
+			newRowSize >>= 1;
+		assert(this->size_ <= newRowSize);
+		this->realloc(newRowSize);
+
+	}
+*/
+public:
+
+	void reset(bool defVal)
+	{
+		std::fill(data_.begin(), data_.end(), defVal);
+	}
+
+	void resize(size_t size, bool defVal = false)
+	{
+
+		if (rowSize_ < size)
+		{
+			this->grow(size, defVal);
+		}
+
+		size_ = size;
+	}
+
+	size_t alloc()
+	{
+		if (size_ >= rowSize_)
+		{
+			this->grow(size_ + 1);
+		}
+
+		return size_++;
+	}
+
+	size_t split(size_t i, bool reflexive = true)
+	{
+		assert(i < size_);
+
+		if (size_ >= rowSize_)
+		{
+			this->grow(size_ + 1);
+		}
+
+		assert((size_ + 1)*rowSize_ <= data_.size());
+
+		// fill collumns
+		auto src = data_.begin() + i;
+		auto end = src + size_*rowSize_;
+		auto dst = data_.begin() + size_;
+
+		for ( ; src != end; src += rowSize_, dst += rowSize_)
+		{
+			*dst = *src;
+		}
+
+		// fill rows
+		src = data_.begin() + i*rowSize_;
+		dst = data_.begin() + size_*rowSize_;
+
+		std::copy(src, src + size_, dst);
+
+		// set the reflexive bit
+		*(dst + size_) = reflexive;
+
+		return size_++;
+	}
+
+	bool get(size_t r, size_t c) const
+	{
+		assert(r < size_ && c < size_);
+		assert(r*rowSize_ + c < data_.size());
+
+		return data_[r*rowSize_ + c];
+	}
+
+	void set(size_t r, size_t c, bool v)
+	{
+		assert(r < size_ && c < size_);
+		assert(r*rowSize_ + c < data_.size());
+
+		data_[r*rowSize_ + c] = v;
+	}
+
+	size_t size() const
+	{
+		return size_;
+	}
+
+public:
+
+	using IndexType    = std::vector<std::vector<size_t>>;
+
+	BinaryRelation(
+		size_t         size = 0,
+		bool           defVal = false,
+		size_t         rowSize = 16) :
+		data_(rowSize*rowSize, defVal),
+		rowSize_(rowSize),
+		size_(0)
+	{
+		this->resize(size, defVal);
+	}
+
+/*
+	BinaryRelation(const BinaryRelation& rel)
+		: data_(rel.data_), row_size_(rel.row_size_), size_(rel.size_) {}
+*/
+
+	BinaryRelation(const std::vector<std::vector<bool> >& rel) :
+		data_(16*16, false),
+		rowSize_(16),
+		size_(0)
+	{
+		this->resize(rel.size(), false);
+		for (size_t i = 0; i < rel.size(); ++i)
+		{
+			assert(rel[i].size() == rel.size());
+			for (size_t j = 0; j < rel.size(); ++j)
+			{
+				this->set(i, j, rel[i][j]);
+			}
+		}
+	}
+
+	// TODO: does this do what is expected? What is expected this should do???
+	// WHAT ABOUT SOME FREAKING DOCUMENTATION??????
```

Comment:
Pure gem :D

Reply:
yes, this is a gem and it should be left here for future generations as a documentation of Ondra's feelings when he studied Jiri's code.